### DISCUSSION TITLE
How to improve forward-declarability of typed actors

### DISCUSSION THREAD

**Discussion by dominiklohmann** [actor-framework/actor-framework#1206] [Q&A] [answered]
## Issue Description

I must forward-declare some of my typed actors so I can assign type IDs to them.

Currently, this works for forward-declaring simple typed actors:

```cpp
#include <caf/fwd.hpp>
#include <caf/replies_to.hpp>

using simple_actor = caf::typed_actor<
  caf::replies_to<add_atom, int, int>::with<int>
>;
```

However, once is-a relationships come into play, the `typed_actor` must be instantiated, which requires another header to be included. This is particularly bad here, because I want my own `fwd.hpp` to be as lean as possible, while also assigning all the required type IDs.

```cpp
#include <caf/fwd.hpp>
#include <caf/replies_to.hpp>

// Expensive! We do not want this in our own fwd.hpp header.
#include <caf/typed_actor.hpp>

using simple_actor = caf::typed_actor<
  caf::replies_to<add_atom, int, int>::with<int>
>;

using complex_actor = caf::typed_actor<
  caf::replies_to<subtract_atom, int, int>::with<int>
>::extend_with<adder_actor>;
```

## Proposed Solution

Instead of requiring instantiation of the `typed_actor` to use `extend_with`, allow for adding other typed actors to the signature list, i.e., make the above `complex_actor` also work like this:

```cpp
using complex_actor = caf::typed_actor<
  simple_actor,
  caf::replies_to<subtract_atom, int, int>::with<int>
>;
```

I think this change should be fairly simple to make. I can work on a patch that does the required changes if desired.

**Comment by Neverlord** [comment]
> However, once is-a relationships come into play

There is no is-a relationship. The handles have set semantics. 🙂

That's why I don't think allowing "nesting" of handles would do any good. Neither conceptually nor in terms of implementing it (would break tons of meta-programming code).

If you don't want to include the typed actor header, here's a solution you can roll on your own:

```c++
template <class Handle>
struct my_wrapper;

template <class... Fs>
struct my_wrapper<caf::typed_actor<Fs...>> {
  template <class... Gs>
  using extend_with = caf::typed_actor<Fs..., Gs...>;
};

using simple_actor = caf::typed_actor<
  caf::result<int>(add_atom, int, int)
>;

using complex_actor = my_wrapper<simple_actor>::extend_with<
  caf::result<int>(subtract_atom, int, int)
>;
```

CAF does the same thing under the hoods.

Just as an aside since you seem to want to minimize included headers: the `replies_to` stuff is no longer necessary with 0.18 and we will deprecate it with the next version. It evaluates to the function signatures I've used above, so you can just type them out yourself and get rid of even more headers and template instantiations. 😉

**Comment by dominiklohmann** [comment]
> > However, once is-a relationships come into play
> 
> There is no is-a relationship. The handles have set semantics. 🙂

Call it a subset-of relationship then, doesn't change the usage. ;-)

> If you don't want to include the typed actor header, here's a solution you can roll on your own:

Thanks for the wrapper, that seems to solve my problem entirely. Great idea, didn't think of doing that.

> Just as an aside since you seem to want to minimize included headers: the `replies_to` stuff is no longer necessary [...]

Thanks for the heads up, will change accordingly. I suppose this can be closed then.

**Comment by dominiklohmann** [comment, marked as answer]
For completeness sake: This is the solution I ended up using.

```cpp
/// Helper utility that enables extending typed actor forwartd declarations
/// without including <caf/typed_actor.hpp>.
template <class... Fs>
struct typed_actor_fwd;

template <class... Fs>
struct typed_actor_fwd {
  template <class Handle>
  struct extend_with_helper;

  template <class... Gs>
  struct extend_with_helper<caf::typed_actor<Gs...>> {
    using type = typed_actor_fwd<Fs..., Gs...>;
  };

  template <class Handle>
  using extend_with = typename extend_with_helper<Handle>::type;

  using unwrap = caf::typed_actor<Fs...>;
};

using c_actor = typed_actor_fwd<
   Handlers...>
  ::extend_with<a_actor>
  ::extend_with<b_actor>
  ::unwrap;
```

@Neverlord I can't move this to Discussions myself, but I feel like this issue would be a good discussion.

**Reply by Neverlord** [reply]
> I feel like this issue would be a good discussion.

ACK. 🙂
